File: linked-list/swap-nodes-in-pairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        ListNode* dummyhead = new ListNode(0);
        dummyhead->next = head;

        ListNode* curr = head;
        ListNode* pre = dummyhead;

        while (curr != nullptr && curr->next != nullptr){
            ListNode* temp = curr->next;
            curr->next = temp->next;

            pre->next = temp;
            temp->next = curr;

            pre = curr;
            curr = pre->next;
        }

        return dummyhead->next;
    }
};

File: recursion/swap-nodes-in-pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummyhead = ListNode()
        dummyhead.next = head
        pre = dummyhead
        curr = head

        while curr and curr.next:
            temp = curr.next
            pre.next = temp
            curr.next = temp.next
            temp.next = curr

            pre = curr
            curr = pre.next

        return dummyhead.next